templates/shift-scheduling/main.go

```diff
@@ -1,370 +1,248 @@
+	// We solve a shift coverage problem using Mixed Integer Programming.
+	// We solve this by generating all possible shifts
+	// and then selecting a subset of these
+	potentialAssignments := []Assignment{}
+	potentialAssignmentsPerDriver := map[int][]Assignment{}
+	for _, driver := range input.Workers {
+		potentialAssignmentsPerDriver[driver.ID] = []Assignment{}
+		for _, availability := range driver.Availability {
+			for start := availability.Start.Time; start.Before(availability.End.Time); start = start.Add(30 * time.Minute) {
+				for end := availability.End.Time; start.Before(end); end = end.Add(-30 * time.Minute) {
+					// make sure that end-start is not more than 8h
+					duration := end.Sub(start)
+					if duration > opts.MaxHoursPerShift {
+						continue
+					}
+					// make sure that end-start is not less than 2h - we are
+					// only shrinking the end time, so we can break here
+					if duration < opts.MinHoursPerShift {
+						break
+					}
+					assignment := Assignment{
+						AssignmentID: len(potentialAssignments),
+						Start:        start,
+						End:          end,
+						Worker:       driver,
+						Duration:     duration,
+					}
+					potentialAssignmentsPerDriver[driver.ID] = append(potentialAssignmentsPerDriver[driver.ID], assignment)
+					potentialAssignments = append(potentialAssignments, assignment)
+				}
+			}
+		}
+	}
 
-	// We define some helper variables that are useful later.
-	nShifts := input.Days * 3
-	workers := model.NewDomain(model.NewRange(1, input.Workers))
-	typeMap := [3]shiftType{"morning", "day", "night"}
-	prefShiftType := map[int]shiftType{}
-	for _, v := range input.Preferences.ShiftType {
-		prefShiftType[int(v.Worker)] = v.Type
+	// initialize demand ids
+	for i, demand := range input.RequiredWorkers {
+		demand.RequiredWorkerID = i
+		input.RequiredWorkers[i] = demand
 	}
 
-	// We track the following variables:
-	// shifts holds one domain for each shift (3 per day in order morning, day,
-	// night) initially it holds all workers available.
-	shifts := store.Repeat(schedule, nShifts, workers)
-
-	// The number of workers active in a schedule.
-	workerCount := store.NewVar(schedule, 0)
-
-	// We also want to maximize worker happiness by fulfilling as many
-	// preferences as possible.
-	happiness := store.NewVar(schedule, 0)
-
-	// As a first step, we can now remove the workers that are unavailable for
-	// certain full days. This results as a new (partial) schedule.
-	changes := []store.Change{}
-
-	for _, v := range input.Unavailable {
-		for _, d := range v.Days {
-			morning := (d - 1) * 3
-			day := morning + 1
-			night := day + 1
-			changes = append(changes,
-				shifts.Remove(morning, []int{int(v.Worker)}),
-				shifts.Remove(day, []int{int(v.Worker)}),
-				shifts.Remove(night, []int{int(v.Worker)}),
-			)
+	demandCovering := map[int][]Assignment{}
+	for _, demand := range input.RequiredWorkers {
```

Comment:
These for loops could also be extracted into a helper function to make `solve` easier to read.

e.g. 
```go
func solve(...) {
  availableShifts := generateAvailableShifts(...)
  demand := generateDemand(...)
  model := buildMIPModel(input, availableShifts, demand)
  ...
}

```